File: TPSServer/packet.py
# coding:utf-8
import struct


class Packet:
    """ 序列化消息 """
    def __init__(self, byte_list=''):
        self.byte_list = byte_list
        self.index = 0
        self.last_length = 0

    def push_byte(self, b):
        self.byte_list += struct.pack('b', b)

    def push_int16(self, num):
        self.byte_list += struct.pack('h', num)

    def push_int32(self, num):
        self.byte_list += struct.pack('i', num)

    def push_int64(self, num):
        self.byte_list += struct.pack('q', num)

    def push_bool(self, flag):
        num = 0
        if flag:
            num = 1
        self.push_byte(num)

    def push_bytes(self, bs):
        self.byte_list += bs

    def push_string(self, string):
        str_len = len(string)
        self.push_int16(str_len)
        self.byte_list += string

    def get_byte(self):
        self.index += self.last_length
        self.last_length = 1
        return struct.unpack('b', self.byte_list[self.index])[0]

    def get_int16(self):
        self.index += self.last_length
        self.last_length = 2
        return struct.unpack('h', self.byte_list[self.index:2])[0]

    def get_int32(self):
        self.index += self.last_length
        self.last_length = 4
        return struct.unpack('i', self.byte_list[self.index:4])[0]

    def get_int64(self):
        self.index += self.last_length
        self.last_length = 8
        return struct.unpack('q', self.byte_list[self.index:8])[0]

    def get_string(self):
        str_len = self.get_int16()
        self.index += self.last_length
        self.last_length = str_len
        return self.byte_list[self.index: self.index+str_len]

    def get_remain(self):
        self.index += self.last_length
        return self.byte_list[self.index:]

File: ThirdPersonShooter/Assets/Script/Network/Packet.cs
using Shaoshuai.Message;
using System;
using System.Collections.Generic;
using System.Text;

/// <summary>
/// 传输协议，包括类型序列化、反序列化操作
/// </summary>
public class Packet
{
    List<Byte> byteList = new List<byte>();
    byte[] bytes;
    int index = 0;
    int lastOffset = 0;
    Int16 strLen = 0;
    byte[] tempBytes;
    public int Length
    {
        get { return bytes.Length - (index + lastOffset); }
    }    
    
    public Packet InitMsg(byte[] bytes)
    {
        this.bytes = bytes;
        index = 0;
        strLen = 0;
        lastOffset = 0;
        return this;
    }

    public byte[] GetByteStream()
    {
        return byteList.ToArray();
    }

    public Int16 getInt16()
    {
        index += lastOffset;
        lastOffset = 2;
        return BitConverter.ToInt16(bytes, index);
    }

    public int getInt32()
    {
        index += lastOffset;
        lastOffset = 4;
        return BitConverter.ToInt32(bytes, index);
    }

    public Int64 getInt64()
    {
        index += lastOffset;
        lastOffset = 8;
        return BitConverter.ToInt64(bytes, index);
    }

    public string getString()
    {
        strLen = getInt16();
        index += lastOffset;
        lastOffset = strLen;
        return Encoding.Unicode.GetString(bytes, index, strLen);
    }

    public byte getByte()
    {
        index += lastOffset;
        lastOffset = 1;
        return bytes[index];
    }

    public bool getBool()
    {
        return getByte() == 1;
    }

    public FixedVec1 getFixedVec1()
    {
        return new FixedVec1(getInt64());
    }

    public FixedVec2 getFixedVec2()
    {
        FixedVec1 v1 = getFixedVec1();
        FixedVec1 v2 = getFixedVec1();
        return new FixedVec2(v1, v2);
    }

    public FixedVec3 getFixedVec3()
    {
        FixedVec1 v1 = getFixedVec1();
        FixedVec1 v2 = getFixedVec1();
        FixedVec1 v3 = getFixedVec1();
        return new FixedVec3(v1, v2, v3);
    }

    public T[] getArray<T>(T[] arr) where T: BaseFormat, new()
    {
        Int16 len = getInt16();
        if (len == 0)
            return null;
        T[] result = new T[len];
        for (int i = 0; i < len; i++)
        {
            if (getBool())
                result[i] = null;
            else
            {
                T value = new T();
                value.Deserialize(this);
                result[i] = value;
            }
        }
        return result;
    }
    //--------------------------------------------------------------------------
    public void push(byte num)
    {
        byteList.Add(num);
    }

    public void push(int num)
    {
        byteList.AddRange(BitConverter.GetBytes(num));
    }

    public void push(Int64 num)
    {
        byteList.AddRange(BitConverter.GetBytes(num));
    }

    public void push(UInt16 num)
    {
        byteList.AddRange(BitConverter.GetBytes(num));
    }

    public void push(bool num)
    {
        byteList.Add(num ? (byte)1: (byte)0);
    }

    public void push(string s)
    {
        tempBytes = Encoding.Unicode.GetBytes(s);
        strLen = (Int16)tempBytes.Length;
        push(strLen);
        byteList.AddRange(tempBytes);
    }

    public void push(FixedVec1 f)
    {
        Int64 v = f.GetValue();
        push(v);
    }
    public void push(FixedVec2 f)
    {
        push(f.x);
        push(f.y);
    }
    public void push(FixedVec3 f)
    {
        push(f.x);
        push(f.y);
        push(f.z);
    }

    public void push<T>(T[] arr) where T: BaseFormat
    {
        ushort len = 0;
        if (arr != null)
            len = (ushort)arr.Length;
        push(len);
        foreach (T a in arr)
        {
            push(a == null);
            if (a != null)
                a.Deserialize(this);
        }
    }
}
